bftengine/tests/testMsgsCertificate/msgsCertificate_test.cpp

```diff
@@ -1,10 +1,132 @@
+const uint16_t selfReplicaId = 0;
+char* reply = new char[10]{};
+int replyLen = strlen(reply);
+
+ReplicaId randomReplicaeId() { return rand() % numOfReplicas; }
+
+class mockCheckpointMsgCmp : public CheckpointInfo {
+ public:
+  static bool equivalent(CheckpointMsg* a, CheckpointMsg* b) {
+    return CheckpointInfo::CheckpointMsgCmp::equivalent(a, b);
+  }
+};
+
+TEST(msgsCertificate_test, obtain_inconsistent_state) {
+  MsgsCertificate<ClientReplyMsg, false, false, true, SimpleClientImp> replysCertificate(
+      numOfReplicas, fval, numOfRequired, selfReplicaId);
+
+  uint8_t reqSeqNum = 0;
+
+  // Add bunch of new messages
+  for (; reqSeqNum < 3; reqSeqNum++) {
+    ReplicaId replicaId = randomReplicaeId();
+    auto newMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+    replysCertificate.addMsg(newMsg, replicaId);
+  }
+
+  ASSERT_TRUE(replysCertificate.isInconsistent());
+
+  replysCertificate.resetAndFree();
+}
+
+TEST(msgsCertificate_test, self_msg_added_peer_msg_ignored) {
+  MsgsCertificate<CheckpointMsg, true, true, true, mockCheckpointMsgCmp> replysCertificate(
+      numOfReplicas, fval, numOfRequired, selfReplicaId);
+
+  auto selfMsg = new CheckpointMsg(selfReplicaId, 0, Digest(), false);
+  replysCertificate.addMsg(selfMsg, selfReplicaId);
+
+  ASSERT_TRUE(replysCertificate.selfMsg() != nullptr);
+
+  replysCertificate.resetAndFree();
+}
+
+TEST(msgsCertificate_test, exceeded_required_msgs_twice) {
+  MsgsCertificate<ClientReplyMsg, false, false, true, SimpleClientImp> replysCertificate(
+      numOfReplicas, fval, numOfRequired, selfReplicaId);
+
+  for (int i = 0; i < 2; i++) {
+    uint8_t numOfReplys = 0;
+    NodeIdType replicaId = 0;
+
+    while (numOfReplys < numOfRequired) {
+      auto msg = new ClientReplyMsg(replicaId, 0, reply, replyLen);
+      replysCertificate.addMsg(msg, replicaId);
+      numOfReplys++;
+      replicaId++;
+    }
+
+    ASSERT_TRUE(replysCertificate.isComplete());
+    replysCertificate.resetAndFree();
+    ASSERT_FALSE(replysCertificate.isComplete());
+  }
+}
+
+TEST(msgsCertificate_test, certifiy_best_correct_msg) {
+  MsgsCertificate<ClientReplyMsg, false, false, true, SimpleClientImp> replysCertificate(
+      numOfReplicas, fval, numOfRequired, selfReplicaId);
+
+  uint8_t reqSeqNum = 0;
+
+  // Create two identical messages and add them.
+
+  ReplicaId replicaId = randomReplicaeId();
+  auto bestMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+  replysCertificate.addMsg(bestMsg, randomReplicaeId());
+
+  replicaId = randomReplicaeId();
+  auto identBestMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+  replysCertificate.addMsg(identBestMsg, randomReplicaeId());
+
+  // Add bunch of new messages
+  for (; reqSeqNum < 4; reqSeqNum++) {
+    replicaId = randomReplicaeId();
+    auto newMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+    replysCertificate.addMsg(newMsg, replicaId);
+  }
```

Comment:
Same thing here, if we draw from RNG, it could happen that some replicaId-s will coincide.